src/bin/test_mpc.rs

```diff
@@ -1,34 +1,97 @@
-use clap::Parser;
-use raw_ipa::{cli::Verbosity, helpers::Role, net::MpcHelperClient};
+use clap::{Parser, Subcommand, ValueEnum};
+use comfy_table::Table;
+use raw_ipa::cli::playbook::{secure_mul, InputSource};
+use raw_ipa::cli::Verbosity;
+use raw_ipa::ff::{Fp31, Fp32BitPrime};
 use std::error::Error;
+use std::fmt::Debug;
+use std::path::PathBuf;
 
 #[derive(Debug, Parser)]
 #[clap(
     name = "mpc-client",
     about = "CLI to execute test scenarios on IPA MPC helpers"
 )]
+#[command(about)]
 struct Args {
     #[clap(flatten)]
     logging: Verbosity,
 
-    #[arg(short, long)]
-    uri: String,
+    #[clap(flatten)]
+    input: CommandInput,
+
+    #[command(subcommand)]
+    action: TestAction,
+}
+
+#[derive(Debug, Parser)]
+pub struct CommandInput {
+    #[arg(
+        long,
+        help = "Read the input from the provided file, instead of standard input"
+    )]
+    input_file: Option<PathBuf>,
+    #[arg(value_enum, long, default_value_t = InputType::Fp32BitPrime, help = "Convert the input into the given field before sending to helpers")]
+    input_type: InputType,
+}
+
+impl From<&CommandInput> for InputSource {
+    fn from(source: &CommandInput) -> Self {
+        if let Some(ref file_name) = source.input_file {
+            InputSource::from_file(file_name)
+        } else {
+            InputSource::from_stdin()
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, ValueEnum)]
+enum InputType {
+    Fp31,
+    Fp32BitPrime,
+    Int64,
+}
+
+#[derive(Debug, Subcommand)]
+enum TestAction {
+    /// Execute end-to-end multiplication.
+    Multiply,
+}
+
+fn print_output<O: Debug>(values: &[Vec<O>; 3]) {
+    let mut shares_table = Table::new();
+    shares_table.set_header(vec!["Row", "H1", "H2", "H3"]);
+    for i in 0..values[0].len() {
+        shares_table.add_row(vec![
+            i.to_string(),
+            format!("{:?}", values[0][i]),
+            format!("{:?}", values[1][i]),
+            format!("{:?}", values[2][i]),
+        ]);
+    }
+
+    println!("{}", shares_table);
 }
 
 #[tokio::main]
 async fn main() -> Result<(), Box<dyn Error>> {
     let args = Args::parse();
     let _handle = args.logging.setup_logging();
 
-    // TODO: Start MPC helpers and discover
-    let client = MpcHelperClient::with_str_addr(&args.uri, Role::H1)?;
-
-    let response = client.echo("hello").await?;
-
-    println!(
-        "Response: {:?}",
-        String::from_utf8_lossy(response.as_slice())
-    );
+    let input = InputSource::from(&args.input);
+    match args.action {
+        TestAction::Multiply => match args.input.input_type {
```

Comment:
I tried VERY hard to make a trait or allow dynamic dispatch here. Ran out of ideas how it can be made possible to do the match on the field type once and call a function with generic argument `F`

Reply:
yea, I've also tried trait approach and almost got it working

```rust
pub trait FieldScenario {
    type Output: Send + Debug;

    fn execute<F>(self, input: &mut InputSource) -> [Vec<Self::Output>; 3]
        where F: Field + IntoShares<Replicated<F>>;
}

pub fn dispatch_field<S, O>(field_type: FieldType, input: &mut InputSource, cl: S) -> [S::Output; 3]
    where S: FieldScenario<Output = O>, O: Send + Debug
{
    match field_type {
        FieldType::Fp31 => FieldScenario::execute::<Fp31>(cl, input),
        FieldType::Fp32BitPrime => FieldScenario::execute::<Fp32BitPrime>(cl, input),
    }
}

impl <O, Fun> FieldScenario for Fun
    where Fun: FnOnce(&mut InputSource) -> [O; 3],
          O: Send + Debug
{
    type Output = O;

    fn execute<F: Field + IntoShares<Replicated<F>>>(self, input: &mut InputSource) -> [Self::Output; 3] {
        self(input)
    }
}

```

I could pass a struct without generic arguments to `dispatch_field` fn, but not the generic function `multiply` parametrized over `F`.